messaging/activity/FirebaseMessagingActivityGenerator.cs

```diff
@@ -0,0 +1,172 @@
+"   * intent does not get set as the Intent that started the app, so when the app comes back online",
+"   * it doesn't see a new FCM message to respond to. As a workaround, we override onNewIntent so",
+"   * that it sends the intent to the MessageForwardingService which forwards the message to the",
+"   * FirebaseMessagingService which in turn sends the message to the application.",
+"   */",
+"  @Override",
+"  protected void onNewIntent(Intent intent) {{",
+"    super.onNewIntent(intent);",
+"",
+"    // If we do not have a 'from' field this intent was not a message and should not be handled. It",
+"    // probably means this intent was fired by tapping on the app icon.",
+"    Bundle extras = intent.getExtras();",
+"    if (extras == null) {{",
+"      return;",
+"    }}",
+"    String from = extras.getString(EXTRA_FROM);",
+"    String messageId = extras.getString(EXTRA_MESSAGE_ID_KEY);",
+"    if (messageId == null) {{",
+"      messageId = extras.getString(EXTRA_MESSAGE_ID_KEY_SERVER);",
+"    }}",
+"    if (from != null && messageId != null) {{",
+"      Intent message = new Intent(this, MessageForwardingService.class);",
+"      message.setAction(MessageForwardingService.ACTION_REMOTE_INTENT);",
+"      message.putExtras(intent);",
+"      message.setData(intent.getData());",
+"      MessageForwardingService.enqueueWork(this, message);",
+"    }}",
+"    setIntent(intent);",
+"  }}",
+"",
+"  /**",
+"   * Dispose of the mUnityPlayer when restarting the app.",
+"   *",
+"   * <p>This ensures that when the app starts up again it does not start with stale data.",
+"   */",
+"  @Override",
+"  protected void onCreate(Bundle savedInstanceState) {{",
+"    if (mUnityPlayer != null) {{",
+"      mUnityPlayer.quit();",
+"      mUnityPlayer = null;",
+"    }}",
+"    super.onCreate(savedInstanceState);",
+"  }}",
+"}}"
+  };
+  private readonly string BaseActivityClass = "UnityPlayerActivity";
+  private readonly string BaseGameActivityClass = "UnityPlayerGameActivity";
+
+  private readonly string GeneratedFileTag = "FirebaseMessagingActivityGenerated";
+  // If this tag is present on the generated file, it will not be replaced.
+  private readonly string PreserveTag = "FirebasePreserve";
+
+  private readonly string OutputPath = Path.Combine("Plugins", "Android");
+  private readonly string OutputFilename = "MessagingUnityPlayerActivity.java";
+
+  public int callbackOrder { get { return 0; } }
+  public void OnPreprocessBuild(BuildReport report) {
+    // Only run this logic when building for Android.
+    if (EditorUserBuildSettings.activeBuildTarget != BuildTarget.Android) {
+      return;
+    }
+
+    // Check if the file has already been generated.
+    string[] oldAssetGuids = AssetDatabase.FindAssets("l:" + GeneratedFileTag);
+    if (oldAssetGuids != null && oldAssetGuids.Length > 0) {
+      if (oldAssetGuids.Length != 1) {
+        Debug.LogWarning("FirebaseMessagingActivityEditor found multiple generated files with the label: " +
```

Comment:
I think the user does not have enough information to act upon this warning message.

I assume this happens when the user move `FirebaseMessagingActivityEditor.java` to a different folder for management purpose.  If there are multiple copies, there could be a risk that the build can fail due to duplicated classes. Here are some potential way I can think of to make this easier for the user:

1. Keep the warning. But list down the path of duplicated files and suggest them to remove the duplication.
2. Change this to `info` level`. 
3. We just remove all the files with this tag if there are more than 2 of them, and regenerate it again. This would force the user to place the file only at the designated location.

Reply:
I believe the only way this would happen would be if the user adds the label to additional files.  If they move it, it should still only find the one file with the label, so that should be fine.  I guess they could potentially copy the file themselves, which might lead to problems.  I can expand on the error message.

Reply:
Yeah. Like you said, there can be two cases:

1. The user accidentally label unrelated assets with this tag, which can be pretty common.  But this is harmless. There is a simple way to rule out this case by using `string.EndWith("MessagingUnityPlayerActivity.java")`
2. If the user copy `MessagingUnityPlayerActivity.java` to a different folder. This case _may_ cause build fail.

I think the minimum we can do is to provide the solution in the log.  Ex. in the end of your log, adds `"Please consider removing duplicated {OutputFilename}, or remove the tag "{PreserveTag}" from unintended assets."`

WDYT?

Reply:
Added a bit more to the log message.